File: sdks/thunkmetrc-ts/wrapper/src/services/AdditivesTemplatesService.ts

import { MetrcClient as InternalMetrcClient } from '@thunkier/thunkmetrc-client';
import { MetrcRateLimiter } from '../RateLimiter';
import { iteratePages } from '../Utils';
import { 
    PaginatedResponse,
    AdditivesTemplate,
    CreateAdditivesTemplatesRequestItem,
    UpdateAdditivesTemplatesRequestItem,
    WriteResponse,
} from '../models';

export class AdditivesTemplatesService {
  constructor(private client: InternalMetrcClient, private rateLimiter: MetrcRateLimiter) {}

  /**
   * Creates new additive templates for a specified Facility.
   * Permissions Required:
   * - Manage Additives
   * POST CreateAdditivesTemplates
   */
  public async createAdditivesTemplates(body: CreateAdditivesTemplatesRequestItem[], licenseNumber?: string): Promise<WriteResponse> {
    return this.rateLimiter.execute(null,false,
        () => this.client.createAdditivesTemplates(body,licenseNumber,
        )
    );
  }

  /**
   * Retrieves a list of active additive templates for a specified Facility.
   * Permissions Required:
   * - Manage Additives
   * GET GetActiveAdditivesTemplates
   */
  public async getActiveAdditivesTemplates(body?: any, lastModifiedEnd?: string, lastModifiedStart?: string, licenseNumber?: string, pageNumber?: number, pageSize?: string): Promise<PaginatedResponse<AdditivesTemplate>> {
    return this.rateLimiter.execute(null,true,
        () => this.client.getActiveAdditivesTemplates(body,lastModifiedEnd,lastModifiedStart,licenseNumber,pageNumber,pageSize,
        )
    );
  }

  /**
   * Retrieves an Additive Template by its Id.
   * Permissions Required:
   * - Manage Additives
   * GET GetAdditivesTemplatesById
   */
  public async getAdditivesTemplatesById(id: string, body?: any, licenseNumber?: string): Promise<AdditivesTemplate> {
    return this.rateLimiter.execute(null,true,
        () => this.client.getAdditivesTemplatesById(id, body,licenseNumber,
        )
    );
  }

  /**
   * Retrieves a list of inactive additive templates for a specified Facility.
   * Permissions Required:
   * - Manage Additives
   * GET GetInactiveAdditivesTemplates
   */
  public async getInactiveAdditivesTemplates(body?: any, licenseNumber?: string, pageNumber?: number, pageSize?: string): Promise<PaginatedResponse<AdditivesTemplate>> {
    return this.rateLimiter.execute(null,true,
        () => this.client.getInactiveAdditivesTemplates(body,licenseNumber,pageNumber,pageSize,
        )
    );
  }

  /**
   * Updates existing additive templates for a specified Facility.
   * Permissions Required:
   * - Manage Additives
   * PUT UpdateAdditivesTemplates
   */
  public async updateAdditivesTemplates(body: UpdateAdditivesTemplatesRequestItem[], licenseNumber?: string): Promise<WriteResponse> {
    return this.rateLimiter.execute(null,false,
        () => this.client.updateAdditivesTemplates(body,licenseNumber,
        )
    );
  }
}

File: sdks/thunkmetrc-rs/wrapper/src/services/additives_templates_service.rs
use thunkmetrc_client::MetrcClient;
use serde_json::Value;
use std::error::Error;
use std::sync::Arc;
#[allow(unused_imports)]
use futures::Stream;
use crate::ratelimiter::MetrcRateLimiter;
#[allow(unused_imports)]
use crate::utils::iterate_pages;
#[allow(unused_imports)]
use crate::models::*;

pub struct AdditivesTemplatesService {
    client: MetrcClient,
    rate_limiter: Arc<MetrcRateLimiter>,
}

impl AdditivesTemplatesService {
    pub fn new(client: MetrcClient, rate_limiter: Arc<MetrcRateLimiter>) -> Self {
        Self {
            client,
            rate_limiter,
        }
    }

    /// POST CreateAdditivesTemplates
    /// Creates new additive templates for a specified Facility.
    /// Permissions Required:
    /// - Manage Additives
    /// Parameters:
    /// - body (Option<&Value>): Request body
    /// - license_number (Option<String>): Filter by licenseNumber
    pub async fn create_additives_templates(&self, license_number: Option<String>, body: Option<Vec<CreateAdditivesTemplatesRequestItem>>) -> std::result::Result<Option<WriteResponse>, Box<dyn Error + Send + Sync>> {
        let body_val = if let Some(b) = body { Some(serde_json::to_value(b)?) } else { None };
        let client = self.client.clone();
        
        let body_val = body_val.clone();

        let resp_val = self.rate_limiter.execute(None,false,
            move || {
                let client = client.clone();
                let license_number = license_number.clone();
                let body_val = body_val.clone();

                async move {
                    client.additives_templates().create_additives_templates(license_number, body_val.as_ref()
                    ).await.map_err(|e| e as Box<dyn Error + Send + Sync>)
                }
            }
        ).await?;

        if let Some(v) = resp_val {
            let typed: WriteResponse = serde_json::from_value(v)?;
            Ok(Some(typed))
        } else {
            Ok(None)
        }
    }

    /// GET GetActiveAdditivesTemplates
    /// Retrieves a list of active additive templates for a specified Facility.
    /// Permissions Required:
    /// - Manage Additives
    /// Parameters:
    /// - last_modified_end (Option<String>): Filter by lastModifiedEnd
    /// - last_modified_start (Option<String>): Filter by lastModifiedStart
    /// - license_number (Option<String>): Filter by licenseNumber
    /// - page_number (Option<String>): Filter by pageNumber
    /// - page_size (Option<String>): Filter by pageSize
    pub async fn get_active_additives_templates(&self, last_modified_end: Option<String>, last_modified_start: Option<String>, license_number: Option<String>, page_number: Option<String>, page_size: Option<String>, body: Option<&Value>) -> std::result::Result<Option<PaginatedResponse<AdditivesTemplate>>, Box<dyn Error + Send + Sync>> {
        let body_val = if let Some(b) = body { Some(serde_json::to_value(b)?) } else { None };
        let client = self.client.clone();
        
        let body_val = body_val.clone();

        let resp_val = self.rate_limiter.execute(None,true,
            move || {
                let client = client.clone();
                let last_modified_end = last_modified_end.clone();
                let last_modified_start = last_modified_start.clone();
                let license_number = license_number.clone();
                let page_number = page_number.clone();
                let page_size = page_size.clone();
                let body_val = body_val.clone();

                async move {
                    client.additives_templates().get_active_additives_templates(last_modified_end, last_modified_start, license_number, page_number, page_size, body_val.as_ref()
                    ).await.map_err(|e| e as Box<dyn Error + Send + Sync>)
                }
            }
        ).await?;

        if let Some(v) = resp_val {
            let typed: PaginatedResponse<AdditivesTemplate> = serde_json::from_value(v)?;
            Ok(Some(typed))
        } else {
            Ok(None)
        }
    }

    /// GET GetAdditivesTemplatesById
    /// Retrieves an Additive Template by its Id.
    /// Permissions Required:
    /// - Manage Additives
    /// Parameters:
    /// - id (str): Path parameter id
    /// - license_number (Option<String>): Filter by licenseNumber
    pub async fn get_additives_templates_by_id(&self, id: &str, license_number: Option<String>, body: Option<&Value>) -> std::result::Result<Option<AdditivesTemplate>, Box<dyn Error + Send + Sync>> {
        let body_val = if let Some(b) = body { Some(serde_json::to_value(b)?) } else { None };
        let id = id.to_string();
        let client = self.client.clone();
        
        let body_val = body_val.clone();

        let resp_val = self.rate_limiter.execute(None,true,
            move || {
                let client = client.clone();
                let id = id.clone();
                let license_number = license_number.clone();
                let body_val = body_val.clone();

                async move {
                    client.additives_templates().get_additives_templates_by_id(&id, license_number, body_val.as_ref()
                    ).await.map_err(|e| e as Box<dyn Error + Send + Sync>)
                }
            }
        ).await?;

        if let Some(v) = resp_val {
            let typed: AdditivesTemplate = serde_json::from_value(v)?;
            Ok(Some(typed))
        } else {
            Ok(None)
        }
    }

    /// GET GetInactiveAdditivesTemplates
    /// Retrieves a list of inactive additive templates for a specified Facility.
    /// Permissions Required:
    /// - Manage Additives
    /// Parameters:
    /// - license_number (Option<String>): Filter by licenseNumber
    /// - page_number (Option<String>): Filter by pageNumber
    /// - page_size (Option<String>): Filter by pageSize
    pub async fn get_inactive_additives_templates(&self, license_number: Option<String>, page_number: Option<String>, page_size: Option<String>, body: Option<&Value>) -> std::result::Result<Option<PaginatedResponse<AdditivesTemplate>>, Box<dyn Error + Send + Sync>> {
        let body_val = if let Some(b) = body { Some(serde_json::to_value(b)?) } else { None };
        let client = self.client.clone();
        
        let body_val = body_val.clone();

        let resp_val = self.rate_limiter.execute(None,true,
            move || {
                let client = client.clone();
                let license_number = license_number.clone();
                let page_number = page_number.clone();
                let page_size = page_size.clone();
                let body_val = body_val.clone();

                async move {
                    client.additives_templates().get_inactive_additives_templates(license_number, page_number, page_size, body_val.as_ref()
                    ).await.map_err(|e| e as Box<dyn Error + Send + Sync>)
                }
            }
        ).await?;

        if let Some(v) = resp_val {
            let typed: PaginatedResponse<AdditivesTemplate> = serde_json::from_value(v)?;
            Ok(Some(typed))
        } else {
            Ok(None)
        }
    }

    /// PUT UpdateAdditivesTemplates
    /// Updates existing additive templates for a specified Facility.
    /// Permissions Required:
    /// - Manage Additives
    /// Parameters:
    /// - body (Option<&Value>): Request body
    /// - license_number (Option<String>): Filter by licenseNumber
    pub async fn update_additives_templates(&self, license_number: Option<String>, body: Option<Vec<UpdateAdditivesTemplatesRequestItem>>) -> std::result::Result<Option<WriteResponse>, Box<dyn Error + Send + Sync>> {
        let body_val = if let Some(b) = body { Some(serde_json::to_value(b)?) } else { None };
        let client = self.client.clone();
        
        let body_val = body_val.clone();

        let resp_val = self.rate_limiter.execute(None,false,
            move || {
                let client = client.clone();
                let license_number = license_number.clone();
                let body_val = body_val.clone();

                async move {
                    client.additives_templates().update_additives_templates(license_number, body_val.as_ref()
                    ).await.map_err(|e| e as Box<dyn Error + Send + Sync>)
                }
            }
        ).await?;

        if let Some(v) = resp_val {
            let typed: WriteResponse = serde_json::from_value(v)?;
            Ok(Some(typed))
        } else {
            Ok(None)
        }
    }
}

